File: app/src/main/java/com/chatai/adr/adapter/ChatAdapter.java
package com.chatai.adr.adapter;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.chatai.adr.R;
import com.chatai.adr.model.ChatMessage;

import java.util.ArrayList;
import java.util.List;

public class ChatAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final List<ChatMessage> messages = new ArrayList<>();
    private final Context context;

    public ChatAdapter(Context context) {
        this.context = context;
    }

    public interface OnMessageActionListener {
        void onSpeak(ChatMessage message);
        void onRegenerate(ChatMessage message, int position);
        void onFeedback(ChatMessage message, boolean isLiked);
    }

    private OnMessageActionListener actionListener;

    public void setOnMessageActionListener(OnMessageActionListener listener) {
        this.actionListener = listener;
    }

    public void setMessages(List<ChatMessage> newMessages) {
        messages.clear();
        if (newMessages != null) {
            messages.addAll(newMessages);
        }
        notifyDataSetChanged();
    }

    public void addMessage(ChatMessage message) {
        messages.add(message);
        notifyItemInserted(messages.size() - 1);
    }

    public void updateMessageContent(int position, String newContent) {
        if (position >= 0 && position < messages.size()) {
            messages.get(position).setContent(newContent);
            notifyItemChanged(position);
        }
    }

    public void clearMessages() {
        int count = messages.size();
        messages.clear();
        notifyItemRangeRemoved(0, count);
    }

    public List<ChatMessage> getMessages() {
        return messages;
    }

    @Override
    public int getItemViewType(int position) {
        return messages.get(position).getSenderType();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        if (viewType == ChatMessage.TYPE_USER) {
            View view = inflater.inflate(R.layout.item_message_user, parent, false);
            return new UserViewHolder(view);
        } else {
            View view = inflater.inflate(R.layout.item_message_ai, parent, false);
            return new AiViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ChatMessage message = messages.get(position);
        if (holder instanceof UserViewHolder) {
            ((UserViewHolder) holder).bind(message);
        } else if (holder instanceof AiViewHolder) {
            ((AiViewHolder) holder).bind(message, position, context, actionListener);
        }
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    static class UserViewHolder extends RecyclerView.ViewHolder {
        private final TextView tvContent;
        private final TextView tvTime;

        UserViewHolder(@NonNull View itemView) {
            super(itemView);
            tvContent = itemView.findViewById(R.id.tvUserContent);
            tvTime = itemView.findViewById(R.id.tvUserTime);
        }

        void bind(ChatMessage message) {
            tvContent.setText(message.getContent());
            tvTime.setText(message.getFormattedTime());
        }
    }

    static class AiViewHolder extends RecyclerView.ViewHolder {
        private final TextView tvContent;
        private final TextView tvTime;
        private final ImageButton btnCopy;
        private final ImageButton btnSpeak;
        private final ImageButton btnRegenerate;
        private final ImageButton btnLike;
        private final ImageButton btnDislike;
        private final LinearLayout layoutCodeBlocks;

        AiViewHolder(@NonNull View itemView) {
            super(itemView);
            tvContent = itemView.findViewById(R.id.tvAiContent);
            tvTime = itemView.findViewById(R.id.tvAiTime);
            btnCopy = itemView.findViewById(R.id.btnCopyAiMessage);
            btnSpeak = itemView.findViewById(R.id.btnSpeakAiMessage);
            btnRegenerate = itemView.findViewById(R.id.btnRegenerateAiMessage);
            btnLike = itemView.findViewById(R.id.btnLikeAiMessage);
            btnDislike = itemView.findViewById(R.id.btnDislikeAiMessage);
            layoutCodeBlocks = itemView.findViewById(R.id.layoutCodeBlocks);
        }

        void bind(ChatMessage message, int position, Context context, OnMessageActionListener listener) {
            String raw = message.getContent() != null ? message.getContent() : "";

            // OpenCode parsing: Check if there are markdown code blocks
            List<com.chatai.adr.util.CodeFormatterUtil.FormattedSection> sections = 
                    com.chatai.adr.util.CodeFormatterUtil.parseContent(raw);

            boolean hasCodeBlock = false;
            for (com.chatai.adr.util.CodeFormatterUtil.FormattedSection sec : sections) {
                if (sec.isCodeBlock) {
                    hasCodeBlock = true;
                    break;
                }
            }

            if (!hasCodeBlock) {
                if (layoutCodeBlocks != null) {
                    layoutCodeBlocks.removeAllViews();
                    layoutCodeBlocks.setVisibility(View.GONE);
                }
                tvContent.setText(com.chatai.adr.util.CodeFormatterUtil.formatInlineMarkdown(raw));
                tvContent.setVisibility(View.VISIBLE);
            } else {
                if (layoutCodeBlocks != null) {
                    layoutCodeBlocks.removeAllViews();
                    layoutCodeBlocks.setVisibility(View.VISIBLE);

                    for (com.chatai.adr.util.CodeFormatterUtil.FormattedSection sec : sections) {
                        if (sec.isCodeBlock) {
                            View codeBlockView = com.chatai.adr.util.CodeFormatterUtil.createCodeBlockView(
                                    context, sec.language, sec.content);
                            layoutCodeBlocks.addView(codeBlockView);
                        } else if (!sec.content.trim().isEmpty()) {
                            TextView inlineTv = new TextView(context);
                            inlineTv.setText(com.chatai.adr.util.CodeFormatterUtil.formatInlineMarkdown(sec.content));
                            inlineTv.setTextColor(context.getResources().getColor(R.color.ai_bubble_text));
                            inlineTv.setTextSize(14.5f);
                            inlineTv.setTextIsSelectable(true);
                            inlineTv.setPadding(0, 8, 0, 8);
                            layoutCodeBlocks.addView(inlineTv);
                        }
                    }
                    tvContent.setVisibility(View.GONE);
                } else {
                    tvContent.setText(com.chatai.adr.util.CodeFormatterUtil.formatInlineMarkdown(raw));
                    tvContent.setVisibility(View.VISIBLE);
                }
            }

            tvTime.setText(message.getFormattedTime());

            // Copy full message
            btnCopy.setOnClickListener(v -> {
                ClipboardManager clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
                ClipData clip = ClipData.newPlainText("AI Message", message.getContent());
                if (clipboard != null) {
                    clipboard.setPrimaryClip(clip);
                    Toast.makeText(context, R.string.copied_to_clipboard, Toast.LENGTH_SHORT).show();
                }
            });

            // Speak (TTS)
            if (btnSpeak != null) {
                btnSpeak.setOnClickListener(v -> {
                    if (listener != null) listener.onSpeak(message);
                });
            }

            // Regenerate
            if (btnRegenerate != null) {
                btnRegenerate.setOnClickListener(v -> {
                    if (listener != null) listener.onRegenerate(message, position);
                });
            }

            // Like
            if (btnLike != null) {
                btnLike.setOnClickListener(v -> {
                    btnLike.setColorFilter(android.graphics.Color.parseColor("#10B981")); // Green
                    if (btnDislike != null) btnDislike.clearColorFilter();
                    Toast.makeText(context, R.string.feedback_liked, Toast.LENGTH_SHORT).show();
                    if (listener != null) listener.onFeedback(message, true);
                });
            }

            // Dislike
            if (btnDislike != null) {
                btnDislike.setOnClickListener(v -> {
                    btnDislike.setColorFilter(android.graphics.Color.parseColor("#EF4444")); // Red
                    if (btnLike != null) btnLike.clearColorFilter();
                    Toast.makeText(context, R.string.feedback_disliked, Toast.LENGTH_SHORT).show();
                    if (listener != null) listener.onFeedback(message, false);
                });
            }
        }
    }
}

File: app/src/main/java/com/chatai/adr/adapter/ChatAdapter.kt
package com.chatai.adr.adapter

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.HorizontalScrollView
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.chatai.adr.R
import com.chatai.adr.model.ChatMessage
import com.chatai.adr.util.CodeFormatterUtil

class ChatAdapter(
    private val messages: MutableList<ChatMessage>,
    private val listener: OnMessageActionListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    interface OnMessageActionListener {
        fun onSpeakText(text: String)
        fun onRegenerate(message: ChatMessage)
    }

    override fun getItemViewType(position: Int): Int {
        val msg = messages[position]
        return if (msg.isUser) VIEW_TYPE_USER else VIEW_TYPE_ASSISTANT
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return if (viewType == VIEW_TYPE_USER) {
            val view = inflater.inflate(R.layout.item_message_user, parent, false)
            UserViewHolder(view)
        } else {
            val view = inflater.inflate(R.layout.item_message_ai, parent, false)
            AssistantViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val message = messages[position]
        if (holder is UserViewHolder) {
            holder.bind(message)
        } else if (holder is AssistantViewHolder) {
            holder.bind(message, listener)
        }
    }

    override fun getItemCount(): Int = messages.size

    fun addMessage(message: ChatMessage) {
        messages.add(message)
        notifyItemInserted(messages.size - 1)
    }

    fun setMessages(newMessages: List<ChatMessage>) {
        messages.clear()
        messages.addAll(newMessages)
        notifyDataSetChanged()
    }

    fun updateLastMessage(content: String, isStreaming: Boolean, modelName: String? = null) {
        if (messages.isNotEmpty()) {
            val lastIdx = messages.size - 1
            val last = messages[lastIdx]
            last.content = content
            last.isStreaming = isStreaming
            if (modelName != null) last.modelName = modelName
            notifyItemChanged(lastIdx)
        }
    }

    class UserViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val tvContent: TextView = itemView.findViewById(R.id.tvUserContent)
        private val tvTime: TextView = itemView.findViewById(R.id.tvUserTime)

        fun bind(message: ChatMessage) {
            tvContent.text = message.content
            tvTime.text = message.formattedTime
        }
    }

    class AssistantViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val tvContent: TextView = itemView.findViewById(R.id.tvAiContent)
        private val layoutCodeBlocks: LinearLayout = itemView.findViewById(R.id.layoutCodeBlocks)
        private val tvTime: TextView = itemView.findViewById(R.id.tvAiTime)
        private val btnCopy: ImageButton = itemView.findViewById(R.id.btnCopyAiMessage)
        private val btnSpeak: ImageButton = itemView.findViewById(R.id.btnSpeakAiMessage)
        private val btnRegenerate: ImageButton = itemView.findViewById(R.id.btnRegenerateAiMessage)

        fun bind(message: ChatMessage, listener: OnMessageActionListener) {
            val context = itemView.context
            tvTime.text = message.formattedTime

            val segments = CodeFormatterUtil.parseSegments(message.content)
            val hasCode = segments.any { it.type == CodeFormatterUtil.SegmentType.CODE }

            if (!hasCode) {
                tvContent.visibility = View.VISIBLE
                tvContent.text = CodeFormatterUtil.formatMarkdown(context, message.content)
                layoutCodeBlocks.visibility = View.GONE
            } else {
                tvContent.visibility = View.GONE
                layoutCodeBlocks.visibility = View.VISIBLE
                layoutCodeBlocks.removeAllViews()

                for (segment in segments) {
                    if (segment.type == CodeFormatterUtil.SegmentType.CODE) {
                        val codeView = createOpenCodeView(context, segment.language, segment.content)
                        layoutCodeBlocks.addView(codeView)
                    } else {
                        val textView = TextView(context).apply {
                            textSize = 14.5f
                            setTextColor(Color.parseColor("#1E293B"))
                            setLineSpacing(4f, 1.15f)
                            setTextIsSelectable(true)
                            text = CodeFormatterUtil.formatMarkdown(context, segment.content)
                            setPadding(0, 4, 0, 4)
                        }
                        layoutCodeBlocks.addView(textView)
                    }
                }
            }

            btnCopy.setOnClickListener {
                val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                val clip = ClipData.newPlainText("ChatAiMessage", message.content)
                clipboard.setPrimaryClip(clip)
                Toast.makeText(context, "Đã sao chép tin nhắn!", Toast.LENGTH_SHORT).show()
            }

            btnSpeak.setOnClickListener {
                listener.onSpeakText(message.content)
            }

            btnRegenerate.setOnClickListener {
                listener.onRegenerate(message)
            }
        }

        private fun createOpenCodeView(context: Context, language: String, code: String): View {
            val card = CardView(context).apply {
                radius = 16f
                cardElevation = 4f
                setCardBackgroundColor(Color.parseColor("#0F172A"))
                val params = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                ).apply {
                    setMargins(0, 10, 0, 10)
                }
                layoutParams = params
            }

            val rootLayout = LinearLayout(context).apply {
                orientation = LinearLayout.VERTICAL
                layoutParams = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
            }

            // Header bar
            val headerBar = LinearLayout(context).apply {
                orientation = LinearLayout.HORIZONTAL
                setBackgroundColor(Color.parseColor("#1E293B"))
                setPadding(16, 10, 16, 10)
            }

            val tvLang = TextView(context).apply {
                text = language
                setTextColor(Color.parseColor("#94A3B8"))
                textSize = 12f
                typeface = Typeface.MONOSPACE
                layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f)
            }

            val tvCopy = TextView(context).apply {
                text = "Sao chép"
                setTextColor(Color.parseColor("#38BDF8"))
                textSize = 12f
                setPadding(12, 4, 12, 4)
                setBackgroundResource(android.R.drawable.list_selector_background)
                setOnClickListener {
                    val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                    val clip = ClipData.newPlainText("Code", code)
                    clipboard.setPrimaryClip(clip)
                    Toast.makeText(context, "Đã chép đoạn mã!", Toast.LENGTH_SHORT).show()
                }
            }

            headerBar.addView(tvLang)
            headerBar.addView(tvCopy)
            rootLayout.addView(headerBar)

            // Code content with horizontal scroll
            val hScroll = HorizontalScrollView(context).apply {
                layoutParams = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
            }

            val tvCode = TextView(context).apply {
                text = code
                setTextColor(Color.parseColor("#E2E8F0"))
                textSize = 13f
                typeface = Typeface.MONOSPACE
                setTextIsSelectable(true)
                setPadding(20, 16, 20, 20)
            }

            hScroll.addView(tvCode)
            rootLayout.addView(hScroll)
            card.addView(rootLayout)

            return card
        }
    }

    companion object {
        private const val VIEW_TYPE_USER = 1
        private const val VIEW_TYPE_ASSISTANT = 2
    }
}
